src/parser.rs

```diff
@@ -40,21 +22,69 @@ named!(
       (((((function-that-returns-another-function arg1 arg2) arg3) arg4) arg5) arg6)
 
    But we'll pretend the first element has to be an operator for now, and that the operands can be recursive. */
-named!(
-    parse_expression_top_level<Expression2>,
-    alt! (
-        parse_num |
-        parse_expression
-    )
-);
+
+fn parse_num2(i: &[u8]) -> IResult<&[u8], Expression2, (&[u8], nom::error::ErrorKind)> {
+    let first_spaces: IResult<&[u8], &[u8], (&[u8], nom::error::ErrorKind)> = take_while(is_space)(i);
```

Comment:
I haven't tested but as these errors seems pretty unlikely (talking about `take_while(is_space)`), lets try to use the [? Operator](https://doc.rust-lang.org/stable/rust-by-example/std/result/question_mark.html) and remove this nasty nesting.

Reply:
Yeah this is definitely gonna be discarded. i think it's not even possible for it to fail. it would either consume the whole string (with a function that always returns true as a param), or none of it, but both outcomes and everything in between is viable.